### PR TITLE
feat(anta): Add support for command blacklist to secure tests

### LINKED ISSUE #413: Feat: Add some security to prevent any "conf" command
TODO:
* Define granularity
* Add CLI switch (later if required)
* By default prevent any modifications - implement at the collect level for tests. ANTA being open source we can only provide a safeguard but not much more.

### PR DESCRIPTION
**Pull request by titom73**
# Description

Add support of a list of blacklisted regex to secure test execution and avoid
any critical change on devices.

It currently blocked CLI with following regex

- Reload command: `^reload\s*\w*`
- Configure mode: `^conf\w*\s*(terminal|session)*`
- Write: `^wr\w*\s*\w+`

Fixes #413 

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have run pre-commit for code linting and typing (`pre-commit run`)
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes (`tox -e testenv`)

### CODE REVIEW

**Comment by gmuloc** [comment]
This implementation won't prevent a malicious device implementation (injecting reload / conf t / ... in its own _collect.

Wondering if we should not rather implement this in the AntaDevice(ABC) collect ?

File: anta/models.py [hunk 1]
```diff
@@ -389,12 +392,26 @@ def render(self, template: AntaTemplate) -> list[AntaCommand]:
...
     async def collect(self) -> None:
         """
         Method used to collect outputs of all commands of this test class from the device of this test instance.
         """
         try:
-            await self.device.collect_commands(self.instance_commands)
+            if self.is_blocked(commands=self.instance_commands) is False:
+                await self.device.collect_commands(self.instance_commands)
+            else:
+                self.result.is_error(message="Test has blocked command")
```
[the remark this answers is not in the document]
Comment (titom73) [reply anta/models.py:414]: Maybe better to state ANTA will protect tests and it would be a best practice to use a r/o user on EOS.

**Review by mtache — CHANGES_REQUESTED** [review]

**Comment by gmuloc** [comment]
> This implementation won't prevent a malicious device implementation (injecting reload / conf t / ... in its own _collect.
> 
> Wondering if we should not rather implement this in the AntaDevice(ABC) collect ?

we keep it in test

File: anta/models.py [hunk 2]
```diff
@@ -33,6 +34,8 @@
 
 DEFAULT_TAG = "all"
 
+BLACKLIST_REGEX = [r"^reload.*", r"^conf\w*\s*(terminal|session)*", r"^wr\w*\s*\w+"]
```
[the remark this answers is not in the document]
Comment (mtache) [reply anta/models.py:38]: You can add the following commands :)

```
clear isis database all
clear isis database level-1
clear isis database level-2
clear isis <instance> database 
```

**Review by mtache — CHANGES_REQUESTED** [review]

File: anta/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtache) [reply anta/models.py:415]: Move line 415 to `blocked` property and add command details in the message.

**Review by mtache — CHANGES_REQUESTED** [review]

**Review by carl-baillargeon — COMMENTED** [review]
Tested a few blocked commands locally with `collect` of `AntaTest` and it works.

Should we add this guard to `anta debug` as well?
```
(.anta) ➜  anta git:(test_blacklist) anta debug run-cmd -c "configure terminal" -d DC1-LEAF1A --ofmt text
Run command configure terminal on DC1-LEAF1A

(.anta) ➜  anta git:(test_blacklist) anta debug run-cmd -c "write" -d DC1-LEAF1A
Run command write on DC1-LEAF1A
{'messages': ['Copy completed successfully.']}
```

**Comment by titom73** [comment]
> Tested a few blocked commands locally with `collect` of `AntaTest` and it works.
> 
> Should we add this guard to `anta debug` as well?
> 
> ```
> (.anta) ➜  anta git:(test_blacklist) anta debug run-cmd -c "configure terminal" -d DC1-LEAF1A --ofmt text
> Run command configure terminal on DC1-LEAF1A
> 
> (.anta) ➜  anta git:(test_blacklist) anta debug run-cmd -c "write" -d DC1-LEAF1A
> Run command write on DC1-LEAF1A
> {'messages': ['Copy completed successfully.']}
> ```

Ideas was to protect tests to send dangerous command while in debug mode, you do that on purpose. A complete protection of the device should be done via a correct RBAC configuration.

Adding this protection to `anta debug` can be done but in different manner than process in place for `anta nrfu`

**Review by carl-baillargeon — APPROVED** [review]
LGTM

**Review by gmuloc — APPROVED** [review]

**Pull request merged**